interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -83,6 +99,61 @@ func (p *Prompt) MarshalJSON() ([]byte, error) {
 	return json.Marshal(toMarshal)
 }
 
+func (p *Prompt) sendReply(outcome prompting.OutcomeType) error {
+	// Reply with any permissions which were previously allowed
+	// If outcome is allow, then reply by allowing all originally-requested
+	// permissions. If outcome is deny, only allow permissions which were
+	// originally requested but have since been allowed by rules, and deny any
+	// remaining permissions.
+	allowedPermission := responseForInterfaceConstraintsOutcome(p.Interface, p.Constraints, outcome)
+	return p.sendReplyWithPermission(allowedPermission)
+}
+
+func responseForInterfaceConstraintsOutcome(iface string, constraints *promptConstraints, outcome prompting.OutcomeType) any {
+	allow, err := outcome.AsBool()
+	if err != nil {
+		// This should not occur, but if so, default to deny
+		allow = false
+		logger.Debugf("%v", err)
+	}
+	allowedPerms := constraints.originalPermissions
+	if !allow {
+		// Remaining permissions were denied, so allow permissions which were
+		// previously allowed by prompt rules
+		allowedPerms = make([]string, 0, len(constraints.originalPermissions)-len(constraints.remainingPermissions))
+		for _, perm := range constraints.originalPermissions {
+			// Exclude any permissions which were remaining at time of denial
+			if !strutil.ListContains(constraints.remainingPermissions, perm) {
+				allowedPerms = append(allowedPerms, perm)
+			}
+		}
+	}
+	allowedPermission, err := prompting.AbstractPermissionsToAppArmorPermissions(iface, allowedPerms)
+	if err != nil {
+		// This should not occur, but if so, permission should be set to the
+		// empty value for its corresponding permission type.
+		logger.Debugf("internal error: cannot convert abstract permissions to AppArmor permissions: %v", err)
+	}
+	return allowedPermission
+}
+
+func (p *Prompt) sendReplyWithPermission(allowedPermission any) error {
+	for _, listenerReq := range p.listenerReqs {
+		if err := sendReply(listenerReq, allowedPermission); err != nil {
+			// Error should only occur if reply is malformed, and since these
+			// listener requests should be identical, if a reply is malformed
+			// for one, it should be malformed for all. Malformed replies should
+			// leave the listener request unchanged. Thus, return early.
+			return err
+		}
+	}
+	return nil
+}
+
+var sendReply = func(listenerReq *listener.Request, allowedPermission any) error {
+	return listenerReq.Reply(allowedPermission)
+}
```

Comment:
if listener comes as the first argument then you can also do:
```suggestion
var sendReply = (*listener.Request).Reply
```